File: crates/grida-text-edit/src/history.rs
use std::time::{Duration, Instant};

use super::TextEditorState;

const DEFAULT_MAX_ENTRIES: usize = 200;
const DEFAULT_MERGE_TIMEOUT: Duration = Duration::from_secs(2);

// ---------------------------------------------------------------------------
// EditKind – classifies text-mutating commands for merge decisions
// ---------------------------------------------------------------------------

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum EditKind {
    Typing,
    Backspace,
    Delete,
    Paste,
    ImeCommit,
    Newline,
}

impl EditKind {
    pub fn is_mergeable(self) -> bool {
        matches!(self, Self::Typing | Self::Backspace | Self::Delete)
    }
}

// ---------------------------------------------------------------------------
// HistoryEntry
// ---------------------------------------------------------------------------

struct HistoryEntry {
    state: TextEditorState,
    kind: EditKind,
    timestamp: Instant,
}

// ---------------------------------------------------------------------------
// EditHistory
// ---------------------------------------------------------------------------

pub struct EditHistory {
    undo_stack: Vec<HistoryEntry>,
    redo_stack: Vec<HistoryEntry>,
    max_entries: usize,
    merge_timeout: Duration,
}

impl EditHistory {
    pub fn new() -> Self {
        Self {
            undo_stack: Vec::new(),
            redo_stack: Vec::new(),
            max_entries: DEFAULT_MAX_ENTRIES,
            merge_timeout: DEFAULT_MERGE_TIMEOUT,
        }
    }

    pub fn can_undo(&self) -> bool {
        !self.undo_stack.is_empty()
    }

    pub fn can_redo(&self) -> bool {
        !self.redo_stack.is_empty()
    }

    pub fn clear(&mut self) {
        self.undo_stack.clear();
        self.redo_stack.clear();
    }

    /// Record the state **before** an edit.
    ///
    /// If the top of the undo stack has the same mergeable kind and the
    /// elapsed time since that entry is within `merge_timeout`, the push is
    /// skipped (we keep the older snapshot so that undo jumps back to the
    /// state before the entire merged run).
    ///
    /// Any pending redo stack is cleared on push.
    pub fn push(&mut self, state_before: &TextEditorState, kind: EditKind) {
        if kind.is_mergeable() {
            if let Some(top) = self.undo_stack.last_mut() {
                if top.kind == kind && top.timestamp.elapsed() < self.merge_timeout {
                    top.timestamp = Instant::now();
                    self.redo_stack.clear();
                    return;
                }
            }
        }

        self.undo_stack.push(HistoryEntry {
            state: state_before.clone(),
            kind,
            timestamp: Instant::now(),
        });

        if self.undo_stack.len() > self.max_entries {
            self.undo_stack.remove(0);
        }

        self.redo_stack.clear();
    }

    /// Undo: saves `current` onto the redo stack and returns the previous state.
    pub fn undo(&mut self, current: &TextEditorState) -> Option<TextEditorState> {
        let entry = self.undo_stack.pop()?;
        self.redo_stack.push(HistoryEntry {
            state: current.clone(),
            kind: entry.kind,
            timestamp: Instant::now(),
        });
        Some(entry.state)
    }

    /// Redo: saves `current` onto the undo stack and returns the next state.
    pub fn redo(&mut self, current: &TextEditorState) -> Option<TextEditorState> {
        let entry = self.redo_stack.pop()?;
        self.undo_stack.push(HistoryEntry {
            state: current.clone(),
            kind: entry.kind,
            timestamp: Instant::now(),
        });
        Some(entry.state)
    }
}

impl Default for EditHistory {
    fn default() -> Self {
        Self::new()
    }
}

// ---------------------------------------------------------------------------
// Test-only helpers
// ---------------------------------------------------------------------------

#[cfg(test)]
impl EditHistory {
    /// Create a history with a custom merge timeout (useful for testing).
    pub fn with_merge_timeout(timeout: Duration) -> Self {
        Self {
            merge_timeout: timeout,
            ..Self::new()
        }
    }

    /// Force-expire the timestamp on the top undo entry so the next push
    /// of the same kind will NOT merge.
    pub fn expire_top(&mut self) {
        if let Some(top) = self.undo_stack.last_mut() {
            top.timestamp -= self.merge_timeout + Duration::from_millis(1);
        }
    }

    pub fn undo_len(&self) -> usize {
        self.undo_stack.len()
    }

    pub fn redo_len(&self) -> usize {
        self.redo_stack.len()
    }

    pub fn with_max_entries(max: usize) -> Self {
        Self {
            max_entries: max,
            ..Self::new()
        }
    }
}

File: packages/grida-text-editor/src/history.ts
/**
 * Session-internal undo/redo stack with merge-by-EditKind grouping.
 *
 * Mirrors `crates/grida/src/text_edit/history.rs`. Same `EditKind`
 * vocabulary, same 2-second merge timeout. Stores snapshots, not
 * diffs — fine at our text lengths.
 *
 * When the orchestrator dispatches a mutation, it classifies the
 * change as an `EditKind`. Mergeable kinds (typing, backspace,
 * delete) within the timeout extend the current run; non-mergeable
 * kinds (paste, ime_commit, cut) always push discrete entries. Style
 * mutations (V2) are also discrete.
 */

import type { EditKind } from "./edit-command";
import type { SessionSnapshot } from "./session";

/**
 * Non-null subset of `EditKind` — every history entry corresponds to a
 * real mutation. `apply_command` returns `EditKind | null`; the
 * orchestrator only pushes when the kind is non-null, so the stack
 * itself never sees `null`.
 */
type RecordedEditKind = Exclude<EditKind, null>;

const MERGE_TIMEOUT_MS = 2000;
/** Hard cap on session-history entries. Prevents unbounded growth
 *  during long edit sessions; the oldest entry is dropped on overflow.
 *  Document-level history (host's HistoryProvider) is the canonical
 *  durable record. */
const MAX_ENTRIES = 100;

const MERGEABLE: ReadonlySet<RecordedEditKind> = new Set<RecordedEditKind>([
  "typing",
  "backspace",
  "delete",
]);

type Entry = {
  before: SessionSnapshot;
  after: SessionSnapshot;
  kind: RecordedEditKind;
  /** ms timestamp of the latest event that contributed to this entry. */
  last_at: number;
};

export class HistoryStack {
  private entries: Entry[] = [];
  /** Index of the next entry to undo. `entries.length` = at head. */
  private cursor = 0;
  /** Test/seam: pluggable clock for deterministic tests. */
  constructor(private readonly now: () => number = () => Date.now()) {}

  get canUndo(): boolean {
    return this.cursor > 0;
  }
  get canRedo(): boolean {
    return this.cursor < this.entries.length;
  }
  get size(): number {
    return this.entries.length;
  }

  /**
   * Record a mutation. `before` is the snapshot *before* the change,
   * `after` is the snapshot *after*. The stack decides whether to
   * merge with the previous entry or push a new one.
   */
  push(
    before: SessionSnapshot,
    after: SessionSnapshot,
    kind: RecordedEditKind
  ): void {
    // Truncate any redo tail.
    if (this.cursor < this.entries.length) {
      this.entries.length = this.cursor;
    }
    const t = this.now();
    const top = this.entries[this.entries.length - 1];
    if (
      top &&
      top.kind === kind &&
      MERGEABLE.has(kind) &&
      t - top.last_at <= MERGE_TIMEOUT_MS
    ) {
      // Merge: keep `before` from the oldest entry (so undo goes back
      // to the start of the run), update `after` and `last_at`.
      top.after = after;
      top.last_at = t;
      return;
    }
    this.entries.push({ before, after, kind, last_at: t });
    if (this.entries.length > MAX_ENTRIES) {
      const drop = this.entries.length - MAX_ENTRIES;
      this.entries.splice(0, drop);
      this.cursor = Math.max(0, this.cursor - drop);
    }
    this.cursor = this.entries.length;
  }

  /** Returns the snapshot to restore, or null if nothing to undo. */
  undo(): SessionSnapshot | null {
    if (!this.canUndo) return null;
    this.cursor--;
    return this.entries[this.cursor].before;
  }

  /** Returns the snapshot to restore, or null if nothing to redo. */
  redo(): SessionSnapshot | null {
    if (!this.canRedo) return null;
    const e = this.entries[this.cursor];
    this.cursor++;
    return e.after;
  }

  clear(): void {
    this.entries = [];
    this.cursor = 0;
  }
}
